### DISCUSSION TITLE
Can we have the UUIDv7 as the rnd_uuid7?

### DISCUSSION THREAD

**Discussion by woss** [questdb/questdb#5267] [Ideas]
We have this -> https://questdb.io/docs/reference/function/random-value-generator/#rnd_uuid4
would it be possible to have uuidv7, one use case is to have it as primary key which is also sortable

**Comment by bluestreak01** [comment]
It would be good to understand the use-case that benefits from uuid7

**Reply by woss** [reply]
For now, I am thinking of using it as a key for a table that stores URL access to resources. i have the unique identifier of the resource which at some user will unpublish and want to merge with the new one, or other resources that represent the resource. 

I can query the records but updating it is not so simple since it doesn't have the unique ID and there are lots of records that have the same resourceID.

I can use the uuid4 or nanoId, hyperID, but i thought having the uuid7 can help me also identify when the record is created ( on the other hand timestamp can do that).

Now when i gave it a thought, maybe the uuid7 doesn't have the use-case.

**Reply by bluestreak01** [reply]
uuid7 is 128-bit value, same as uuid4, it should be a matter of writing a new `rndUuid7()`, which is trivial, may be you can smash in the contribution? Based on `rndUuid4()`?

**Reply by jerrinot** [reply]
There is a minor complication with UUID v7 (compared to UUID v4): the time component should use the clock from `SqlExecutionContext.getNow()` to maintain deterministic generation. Determinism is important for WAL replay, such as on database replicas. 

However, consider this query: `insert into tab (select rnd_uuidv7() from long_sequence(100_000_000))`.
This query will generate 100M UUIDs. You probably don't want all of them to have the same timestamp, but that's exactly what `SqlExecutionContext.getNow()` provides. This creates two potential issues:
1. The UUIDs lose their effectiveness as time-ordered identifiers since they all share the same temporal component
2. The probability of collisions increases significantly because we're relying solely on the random bits (~74 bits) for uniqueness, rather than also having the benefit of an incrementing 48-bit timestamp component.

It's probably not a show-stopper, but it's something to consider when implementing UUIDv7 generator function.

**Reply by bluestreak01** [reply]
now() is a result-set stable function on purpose. `rnd_uuit7()` can just get micros
